### PR TITLE
Implementation of recovering public polynomial

### PR DESCRIPTION
**Pull request by nikkolasg**
This PR implements the recovering of a public polynomial (ie. all its coefficients) from a list of public shares. This is needed for the newest features coming into DKG.

### CODE REVIEW

**Comment by Daeinar** [comment]
@nikkolasg Looks good in general, I only have one small question: https://github.com/dedis/kyber/pull/322/files#r215196152

File: share/poly.go [hunk 1]
```diff
@@ -426,3 +426,66 @@ func RecoverCommit(g kyber.Group, shares []*PubShare, t, n int) (kyber.Point, er
...
+	}
+
+	if len(x) < t {
+		return nil, errors.New("share: not enough good public shares to reconstruct secret commitment")
+	}
+
+	var accPoly *PubPoly
+	var err error
+
+	num := g.Scalar()
```
Comment (Daeinar) [inline share/poly.go]: `num` seems to be never used other than being set to 1 in line 457?

**Comment by Daeinar** [comment]
Modulo the two smaller comments above this looks good to me.

File: share/poly.go [hunk 2]
```diff
@@ -426,3 +410,69 @@ func RecoverCommit(g kyber.Group, shares []*PubShare, t, n int) (kyber.Point, er
...
+	var err error
+
+	for j := range x {
+		basis := LagrangeBasis(g, j, x)
+		// compute the L_j * y_j polynomial in point space
+		tmp := basis.Commit(shares[j].V)
+		if accPoly == nil {
+			accPoly = tmp
+			continue
+		}
```
Comment (jeffallen) [inline share/poly.go]: I prefer to read loops which initialize an accumulator when they are wrtitten with an else and no continue:
```
for {
  if acc == nil {
    acc = tmp
  } else {
   var err error
    acc, err = add.Add(tmp)
    if err != nil {
      return nil, err
    }
  }
} //end for
```

File: share/poly.go [hunk 3]
```diff
@@ -234,29 +234,13 @@ func RecoverPriPoly(g kyber.Group, shares []*PriShare, t, n int) (*PriPoly, erro
 
 	var accPoly *PriPoly
 	var err error
-	den := g.Scalar()
+	//den := g.Scalar()
```
Comment (jeffallen) [inline share/poly.go]: Please clean this up.

File: share/poly.go [hunk 4]
```diff
@@ -426,3 +410,69 @@ func RecoverCommit(g kyber.Group, shares []*PubShare, t, n int) (kyber.Point, er
...
+	}
+
+	return accPoly, nil
+
+}
+
+// LagrangeBasis returns a PriPoly containing the Lagrange coefficients for the
+// i-th position. xs is a mapping between the indices and the values that the
+// interpolation is using, computed with xScalar().
+func LagrangeBasis(g kyber.Group, i int, xs map[int]kyber.Scalar) *PriPoly {
```
Comment (jeffallen) [inline share/poly.go]: Is this function useful on it's own? Or should it be lower-case, so not exported?

Comment (nikkolasg) [reply share/poly.go]: Good point, it'll be enough if we don't export it. I'll do that.

File: share/poly.go (hunk 2 above)
Comment (nikkolasg) [reply share/poly.go]: Mhhhh. I understand your preference, mine is more along the lines of clarity of code: 
+ It's simpler / easier to read without an additional indentation and if/else case to mentally remember.
+ It clearly indicates a breaking condition in the loop, same as returning early from a function.  

For example, we never do this:
```go
if err := myDangerousOperation(); err != nil {
   return err
} else {
   mySafeOperation()
}
```
but simply
```go
if err := myDangerousOperation(); err != nil {
   return err
} 
mySafeOperation()
```
What's your reasonings for not liking `continue` ? C/C++ background ? ;)

**Pull request merged**